src/support/scale.rs

```diff
@@ -0,0 +1,319 @@
+//! Support for the [`parity-scale-codec`](https://crates.io/crates/parity-scale-codec) crate.
+#![cfg(feature = "parity-scale-codec")]
+#![cfg_attr(has_doc_cfg, doc(cfg(feature = "parity-scale-codec")))]
+
+use crate::Uint;
+use parity_scale_codec::{
+    Compact, CompactAs, Decode, Encode, EncodeAsRef, EncodeLike, Error, HasCompact, Input,
+    MaxEncodedLen, Output,
+};
+
+// Compact encoding is supported only for 0-(2**536-1) values:
+// https://docs.substrate.io/reference/scale-codec/#fn-1
+pub(crate) const COMPACT_BITS_LIMIT: usize = 536;
+
+impl<const BITS: usize, const LIMBS: usize> Encode for Uint<BITS, LIMBS> {
+    /// u32 prefix for compact encoding + bytes needed for LE bytes representation
+    fn size_hint(&self) -> usize {
+        core::mem::size_of::<u32>() + Self::BYTES
+    }
+
+    fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+        self.as_le_bytes().using_encoded(f)
+    }
+}
+
+impl<const BITS: usize, const LIMBS: usize> MaxEncodedLen for Uint<BITS, LIMBS> {
+    fn max_encoded_len() -> usize {
+        core::mem::size_of::<Self>()
+    }
+}
+
+impl<const BITS: usize, const LIMBS: usize> Decode for Uint<BITS, LIMBS> {
+    fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+        Decode::decode(input).and_then(|b: Vec<_>| {
+            Self::try_from_le_slice(&b).ok_or(Error::from("value is larger than fits the Uint"))
+        })
+    }
+}
+
+// TODO: Use nightly generic const expressions to validate that BITS parameter is less than 536
```

Comment:
A workaround is to add `assert!(BITS < 536)`s in the constructs. Since this is const it will optimize away if everything is good. If the users makes a mistake, it will panic the moment a type is constructed. Presumably the users will at least construct the type during testing.

Reply:
Ah, you have this with `assert_compact_supported`, I would add it to the `From` construtors too.